File: linked-list/odd-even-linked-list.js
// 328. Odd Even Linked List
// https://leetcode.com/problems/odd-even-linked-list/description/

var oddEvenList = function (head) {
  if (!head) return head;

  var odd = head;
  var even = head.next;

  while (odd.next && odd.next.next) {
    var tmp = odd.next;
    odd.next = odd.next.next;
    odd = odd.next;
    tmp.next = odd.next;
  }
  odd.next = even;
  return head;
};

File: linked-list/odd-even-linked-list.java
// https://leetcode.com/problems/odd-even-linked-list/description/

// Java
class Solution {
    public ListNode oddEvenList(ListNode head) {

        if (head == null || head.next == null) return  head;

        ListNode odd = head, even = head.next, oddHead = head, evenHead = head.next;

        // use odd as a condition because if odd doesnt exit, even cant either
        // odd.next implies even, odd.next.next implies even.next
        while (odd.next != null && odd.next.next != null){
            odd.next = odd.next.next;
            even.next = even.next.next;
            odd = odd.next;
            even = even.next;
        }

        odd.next = evenHead;

        return oddHead;
    }
}

// head will be set as odd, head.next will be set as even

// odd.next = odd.next.next
// even.next = even.next.next
// odd = odd.next
// even = even.next